st2actions/st2actions/scheduler/handler.py

```diff
@@ -0,0 +1,215 @@
+from st2common.constants import action as action_constants
+from st2common.exceptions.db import StackStormDBObjectNotFoundError
+from st2common.models.db.liveaction import LiveActionDB
+from st2common.services import action as action_service
+from st2common.services import policies as policy_service
+from st2common.persistence.liveaction import LiveAction
+from st2common.persistence.execution_queue import ExecutionQueue
+from st2common.util import action_db as action_utils
+from st2common.metrics import base as metrics
+from st2common.exceptions import db as db_exc
+
+__all__ = [
+    'ExecutionQueueHandler',
+    'get_handler'
+]
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ExecutionQueueHandler(object):
+    def __init__(self):
+        self.message_type = LiveActionDB
+        self._shutdown = False
+        self._pool = eventlet.GreenPool(size=10)
+
+    def cleanup(self):
+        LOG.debug('Starting scheduler garbage collection')
+
+        while self._shutdown is not True:
+            eventlet.greenthread.sleep(5)
+            self._handle_garbage_collection()
+
+    def run(self):
+        LOG.debug('Entering scheduler loop')
+
+        while self._shutdown is not True:
+            eventlet.greenthread.sleep(0.1)
+
+            with metrics.Timer(key='scheduler.loop'):
+                execution = self._get_next_execution()
+
+            with metrics.Timer(key='scheduler.loop.spawn_execution'):
+                if execution:
+                    self._pool.spawn(self._handle_execution, execution)
+
+    def _handle_garbage_collection(self):
+        query = {
+            "scheduled_start_timestamp__lte": date.append_milliseconds_to_time(
+                date.get_datetime_utc_now(),
+                -60000
+            ),
+            "handling": True,
+        }
+
+        executions = ExecutionQueue.query(**query)
+
+        if executions:
+            for execution in executions:
+                execution.handling = False
+
+                try:
+                    ExecutionQueue.add_or_update(execution, publish=False)
+                    LOG.info("Removing lock for orphaned execution: %s", execution.id)
+                except db_exc.StackStormDBObjectWriteConflictError:
+                    LOG.info(
+                        "Execution updated before rescheduling: %s",
+                        execution.id
+                    )
+
+    def _get_next_execution(self):
+        """
+            Sort executions by fifo and priority and get the latest, highest priority
+            item from the queue and pop it off.
+        """
+        query = {
+            "scheduled_start_timestamp__lte": date.get_datetime_utc_now(),
+            "handling": False,
+            "limit": 1,
+            "order_by": [
+                "scheduled_start_timestamp",
+            ]
+        }
+
+        execution = ExecutionQueue.query(**query).first()
+
+        if execution:
+            execution.handling = True
```

Comment:
Unless I'm missing something - it looks like there could still be race if we run multiple ``st2scheduler`` processes since we don't perform any locking here.

Or am I missing something?

Reply:
EDIT: Never mind I see it's using ``ChangeRevisionMongoDBAccess`` which should prevent that.

Would still be good to add a comment to this function on how coordination is handled since it's not immediately obvious from the code.